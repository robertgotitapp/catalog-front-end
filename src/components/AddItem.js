import React, { Component } from 'react';
import { connect } from 'react-redux';
import Form from 'react-bootstrap/Form';
import Button from 'react-bootstrap/Button';
import { addItem } from '../actions/items';

export class AddItem extends Component {
    state = {
      name: '',
      price: 0,
      description: '',
    }

    handleChange = (e) => {
      this.setState({
        [e.target.name]: e.target.value,
      });
    }

    handleSubmit = (e) => {
      e.preventDefault();
      this.props.addItem(this.props.categoryId, { ...this.state });
    }

    render() {
      const { name, price, description } = this.state;

      return (
        <div>
          <Form onSubmit={this.handleSubmit}>
            <Form.Group>
              <Form.Label>Item Name</Form.Label>
              <Form.Control type="text" name="name" onChange={this.handleChange} value={name} />
            </Form.Group>
            <Form.Group>
              <Form.Label>Item Price</Form.Label>
              <Form.Control type="number" name="price" onChange={this.handleChange} value={price} />
            </Form.Group>
            <Form.Group>
              <Form.Label>Item Description</Form.Label>
              <Form.Control type="text" name="description" onChange={this.handleChange} value={description} />
            </Form.Group>
            <Button variant="primary" type="submit">
              Submit
            </Button>
          </Form>
        </div>
      );
    }
}

function mapStateToProps({ usersReducer }) {
  return {
    users: usersReducer,
  };
}

const mapDispatchtoProps = {
  addItem,
};

export default connect(mapStateToProps, mapDispatchtoProps)(AddItem);
