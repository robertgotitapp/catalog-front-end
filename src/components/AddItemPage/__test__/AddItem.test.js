import React from 'react';
import { shallow } from 'enzyme';
import Alert from 'react-bootstrap/Alert';
import { AddItem } from '../AddItem';

describe('components/AddItem', () => {
  let wrapper;
  let props;

  beforeEach(() => {
    props = {
      categoryIds: [1, 2, 3, 4],
      categories: {
        1: {
          created: '2019-06-17T13:46:18+00:00',
          description: 'Laptop',
          id: 1,
          name: 'Laptop',
          updated: '2019-06-17T13:46:18+00:00',
        },
        2: {
          created: '2019-06-17T13:46:25+00:00',
          description: 'Books',
          id: 2,
          name: 'Books',
          updated: '2019-06-17T13:46:25+00:00',
        },
        3: {
          created: '2019-06-17T13:46:48+00:00',
          description: 'Speaker',
          id: 3,
          name: 'Speaker',
          updated: '2019-06-17T13:46:48+00:00',
        },
        4: {
          created: '2019-06-17T13:47:17+00:00',
          description: 'Vehicle',
          id: 4,
          name: 'Vehicle',
          updated: '2019-06-17T13:47:17+00:00',
        },
      },
      addItem: jest.fn(() => Promise.resolve({
        statusCode: 1,
      })),
      history: { push: jest.fn() },
    };
  });

  const setup = () => {
    wrapper = shallow(<AddItem {...props} />);
  };

  const inputValidItemDetail = () => {
    wrapper.find('[name="name"]')
      .simulate('change',
        { target: { name: 'name', value: 'Test Item 1' } });

    wrapper.find('[name="description"]')
      .simulate('change',
        { target: { name: 'description', value: 'Test Item 1' } });

    wrapper.find('[name="price"]')
      .simulate('change',
        { target: { name: 'price', value: 10 } });
  };

  const inputInvalidItemDetail = () => {
    wrapper.find('[name="name"]')
      .simulate('change',
        { target: { name: 'name', value: 'Test' } });

    wrapper.find('[name="description"]')
      .simulate('change',
        {
          target: {
            name: 'description',
            value: 'rojt9QGe5MP7AXXYzi9SmREjtBkLg9OTMUHO9WKkujABJPbx2gAcNC'
            + 'y11WkuXZ52dPzHE95m72q1HOeAcfQkAQ7gJbh9kiWBSac7OnW4Vu2hrpkGYk1T1gVJoeBD'
            + 'LLFO3GsB5uo9YYQOiURst6MEXkS3jW2aIBEF2tYIabAJups0nDb3M7y8fExN7M4EqriAxR4cctrItgi5te72e9IHj7zYvh8y',
          },
        });

    wrapper.find('[name="price"]')
      .simulate('change',
        { target: { name: 'price', value: -10 } });
  };

  const submitForm = () => {
    wrapper.instance().handleSubmit({
      preventDefault: jest.fn(),
    });
  };

  it('should render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should trigger state change when FormControl name got input', () => {
    setup();
    const itemNameInput = wrapper.find('[name="name"]');
    itemNameInput.simulate('change', { target: { name: 'name', value: 'item 1' } });
    expect(wrapper.state().name).toEqual('item 1');
  });

  it('should trigger state change when FormControl description got input', () => {
    setup();
    const itemDescriptionInput = wrapper.find('[name="description"]');
    itemDescriptionInput.simulate('change', { target: { name: 'description', value: 'description' } });
    expect(wrapper.state().description).toEqual('description');
  });

  it('should trigger state change when FormControl price got input', () => {
    setup();
    const itemPriceInput = wrapper.find('[name="price"]');
    itemPriceInput.simulate('change', { target: { name: 'price', value: 100 } });
    expect(wrapper.state().price).toEqual(100);
  });

  it('should trigger state change when selectCategory function got input', () => {
    setup();
    wrapper.instance().selectCategory({
      preventDefault: jest.fn(),
      target: { name: 'selectCategory', value: 2 },
    });
    expect(wrapper.state().selectedCategory).toEqual(2);
  });

  it('should trigger history push function when valid item detail is submitted', async () => {
    setup();
    inputValidItemDetail();
    submitForm();
    await Promise.resolve();
    expect(props.history.push).toBeCalled();
  });


  it('should render Alert when invalid item detail is submitted', async () => {
    setup();
    inputInvalidItemDetail();
    submitForm();
    expect(wrapper.find(Alert)).toHaveLength(3);
  });

  it('should display alerts if the request to api is failed', async () => {
    props = {
      categoryIds: [1, 2, 3, 4],
      categories: {
        1: {
          created: '2019-06-17T13:46:18+00:00',
          description: 'Laptop',
          id: 1,
          name: 'Laptop',
          updated: '2019-06-17T13:46:18+00:00',
        },
        2: {
          created: '2019-06-17T13:46:25+00:00',
          description: 'Books',
          id: 2,
          name: 'Books',
          updated: '2019-06-17T13:46:25+00:00',
        },
        3: {
          created: '2019-06-17T13:46:48+00:00',
          description: 'Speaker',
          id: 3,
          name: 'Speaker',
          updated: '2019-06-17T13:46:48+00:00',
        },
        4: {
          created: '2019-06-17T13:47:17+00:00',
          description: 'Vehicle',
          id: 4,
          name: 'Vehicle',
          updated: '2019-06-17T13:47:17+00:00',
        },
      },
      addItem: jest.fn(() => Promise.resolve({
        statusCode: 0,
        errors: {
          message: {
            name: 'Item name has to be at least 5 characters',
          },
        },
      })),
    };
    setup();
    inputValidItemDetail();
    submitForm();
    await Promise.resolve();
    expect(wrapper.find(Alert)).toHaveLength(1);
  });
});
