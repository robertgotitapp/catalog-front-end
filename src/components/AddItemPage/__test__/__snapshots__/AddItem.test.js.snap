// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/AddItem should render correctly 1`] = `
<div>
  <h2>
    Add Item
  </h2>
  <Form
    inline={false}
    onSubmit={[Function]}
  >
    <FormGroup>
      <FormLabel
        column={false}
        srOnly={false}
      >
        Item Name
      </FormLabel>
      <FormControl
        name="name"
        onChange={[Function]}
        type="text"
        value=""
      />
    </FormGroup>
    <FormGroup>
      <FormLabel
        column={false}
        srOnly={false}
      >
        Item Price
      </FormLabel>
      <FormControl
        name="price"
        onChange={[Function]}
        type="number"
        value={0}
      />
    </FormGroup>
    <FormGroup>
      <FormLabel
        column={false}
        srOnly={false}
      >
        Item Description
      </FormLabel>
      <FormControl
        name="description"
        onChange={[Function]}
        type="text"
        value=""
      />
    </FormGroup>
    <FormControl
      as="select"
      name="selectedCategory"
      onChange={[Function]}
    >
      <option
        key="1"
        value={1}
      >
        Laptop
      </option>
      <option
        key="2"
        value={2}
      >
        Books
      </option>
      <option
        key="3"
        value={3}
      >
        Speaker
      </option>
      <option
        key="4"
        value={4}
      >
        Vehicle
      </option>
    </FormControl>
    <Button
      active={false}
      disabled={false}
      type="submit"
      variant="primary"
    >
      Submit
    </Button>
  </Form>
</div>
`;
